rust/lance/src/index/vector/pq.rs

```diff
@@ -130,65 +129,63 @@ impl PQIndex {
         }))
     }
 
-    fn cosine_distances(&self, key: &Float32Array, code: &UInt8Array) -> Result<ArrayRef> {
+    /// Pre-compute cosine distance to each sub-centroids.
+    ///
+    /// Parameters
+    ///  - query: the query vector, with shape (dimension, )
+    ///  - code: the PQ code in one partition.
+    ///
+    fn cosine_distances(&self, query: &Float32Array, code: &UInt8Array) -> Result<ArrayRef> {
         // Build two tables for cosine distance.
         //
         // xy table: `[f32: num_sub_vectors(row) * num_centroids(column)]`.
         // y_norm table: `[f32: num_sub_vectors(row) * num_centroids(column)]`.
-        let mut xy_table: Vec<f32> = vec![];
-        let mut y_norm_table: Vec<f32> = vec![];
-
-        let x_norm = norm_l2(key.values()).powi(2);
+        let num_centroids = ProductQuantizer::num_centroids(self.nbits);
+        let mut xy_table: Vec<f32> = Vec::with_capacity(self.num_sub_vectors * num_centroids);
+        let mut y2_table: Vec<f32> = Vec::with_capacity(self.num_sub_vectors * num_centroids);
 
+        let x_norm = norm_l2(query.values());
         let sub_vector_length = self.dimension / self.num_sub_vectors;
         for i in 0..self.num_sub_vectors {
-            let key_sub_vector: Float32Array = key.slice(i * sub_vector_length, sub_vector_length);
+            // The sub-vector section of the query vector.
+            let key_sub_vector =
+                &query.values()[i * sub_vector_length..(i + 1) * sub_vector_length];
             let sub_vector_centroids = self.pq.centroids(i).ok_or_else(|| Error::Index {
                 message: "PQIndex::cosine_distances: PQ is not initialized".to_string(),
             })?;
             let xy = sub_vector_centroids
-                .as_ref()
                 .values()
                 .chunks_exact(sub_vector_length)
-                .map(|cent| {
-                    // Accelerate this later.
-                    cent.iter()
-                        .zip(key_sub_vector.values().iter())
-                        .map(|(y, x)| (x - y).powi(2))
-                        .sum::<f32>()
-                });
+                .map(|cent| cent.dot(key_sub_vector));
             xy_table.extend(xy);
 
             let y_norm = sub_vector_centroids
-                .as_ref()
                 .values()
                 .chunks_exact(sub_vector_length)
-                .map(|cent| {
-                    // Accelerate this later.
-                    cent.iter().map(|y| y.powi(2)).sum::<f32>()
-                });
-            y_norm_table.extend(y_norm);
+                .map(|cent| cent.norm_l2().powf(2.0));
```

Comment:
Can you drop the `.powf(2.0)` here and the `.sqrt` of `y_norm.sqrt()` below?  Not a correctness thing, just a minor perf suggestion.

Reply:
We can not. 

so this is an array of `y^2`.    

$`\sum{\sqrt{y^2}} != \sum_{0}^{n}{\sqrt{y_i^2}}`$

Reply:
It just seems like the two operations will cancel each other out. If you remove all the middle code you basically have...

```
let y_norm_for_cent = cent.norm_l2().powf(2.0);
let cosine_dist = `1.0 - xy / (x_norm_for_centroid * y_norm_for_centroid.sqrt())
```

So the right term is basically: `cent.norm_l2().powf(2.0).sqrt()` which should be identical to `cent.norm_l2()` right?

Reply:
y_norm_table build a table of y^2 for each sub-vector. 
the last `y_norm_for_centroid` is the sum of these y^2 for each PQ subvector tho.